### ISSUE TITLE
New syntax for specifying enum element type

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#5417]
Today, `enum` element types other than `int(64)` must be specified on one of the elements like this:

```chapel
enum E {a = 1 : int(32), b, c};
```

It would be nice to be able to specify the element type outside of the constant list, e.g.

```chapel
enum E: int(32) {a, b, c};
```

**Comment by bradcray** [comment]
@ben-albrecht : Looking into #8552, I think your assertion above that casting the integer value of an enum sets the size of that enum in some way is incorrect (or at least is no longer correct).  It definitely read as a complete surprise to me, and the code doesn't seem to be backing up that claim.  Do you know where you got this idea?

**Comment by ben-albrecht** [comment]
>Today, `enum` element types other than `int(64)` must be specified on one of the elements like this:
>
>```chapel
>enum E {a = 1 : int(32), b, c};
>```

I checked out a commit (`8dc84bd`) from around the time this was filed and confirmed that the original example did not set the size of the enum as I had asserted. My confirmation of the type must have been flawed in some way. Sorry to cause confusion. I'll edit the original description.

**Comment by bradcray** [comment]
In retrospect, the `=` in the proposal you attribute to me above was probably a think-o on my part (if we don't use / require `=` in the general case, I think it'd be confusing to do so here).  But just to put a second option on the table, another way we could specify it might be:

```chapel
enum E {a, b, c}: int(32);
```

and perhaps this would be slightly preferable by making the specification more of an afterthought or appendix and less in the way...

**Comment by bradcray** [comment]
As we've been working on revamping and improving enums, we've removed implicit casts from enums to integers which suggests that all these declarations would really need now is a bit-width for how many bits to devote to the representation of the enum.  (Since any conversion from enum to int will now need to specify what int type/size to convert it to).  That leads me to think about syntaxes like:

```chapel
enum(32) E {a, b, c};
enum E(32) {a, b, c};
enum E {a, b, c}(32);
```

I'm not wildly crazy about any of these, but gravitate toward the second a bit more than the other two.  Open to other suggestions as well.

**Comment by ben-albrecht** [comment]
Despite only needing the bit-width, would it be that bad to still require a full type as you originally envisioned?

```chpl
enum E {a, b, c}: int(32);
```

That seems more natural to me than the examples listed above.

**Comment by bradcray** [comment]
Perhaps.  The only thing that seems _slightly_ odd about that to me is that I can't think of a way in which `int(32)` and `uint(32)` would differ from one another if used in this position.  But as you say, maybe that doesn't really matter if it looks sufficiently cleaner (and may lead a user who wasn't familiar with the capability to guess it's purpose more quickly than the alternatives).

**Comment by ben-albrecht** [comment]
> The only thing that seems _slightly_ odd about that to me is that I can't think of a way in which `int(32)` and `uint(32)` would differ from one another if used in this position. 

Would it make sense for a user to specify a type of `uint(32)` for the enum?

**Comment by bradcray** [comment]
> Would it make sense for a user to specify a type of uint(32) for the enum?

I don't think it would _not_ make sense, but I'm not sure what good it would do (in the sense, that casts would still be required to convert an enum const to either `uint(32)` or any other integral type).  I suppose it could be used to force an error if any of the enum's symbols were given a negative underlying integer value?

**Comment by ben-albrecht** [comment]
If it didn't make sense, it could be a compiler error. If it had a use, then it could be allowed.

Anyway, that's just one idea. We should get input from others.

**Comment by leekillough** [comment]
I immediately thought about a feature like this while reviewing https://github.com/chapel-lang/chapel/pull/17481.

In C++ terminology, it's called the _[underlying type](https://en.cppreference.com/w/cpp/types/underlying_type)_ of an `enum`.

I think either signed or unsigned `int` should be okay. If you have an `enum` with 129 values, `uint(8)` makes more sense than `int(8)`, for example.

The C++ [syntax](https://en.cppreference.com/w/cpp/language/enum) is `enum [ class | struct ] [ attr ] enum-name [ : enum-base ] [ { enumeration-list } ];`

**Comment by e-kayrakli** [comment]
Today, I ran into an issue where I would have liked to have a smaller enum. In my case, I have an array of enums. I pass that array to an extern C function. I was running into issues because the C function's formal was of type `int*` (32-bit values). So, in C, I was iterating over 32-bit boundaries of this array, which was getting junk values. This is clearly an error on my part, but my first thought to fix it was to make my enum smaller. For this particular case, making the formal on the C function to be 64-bits is not a big deal in my case. So, this is not a blocker for me.
